### PR TITLE
move logic about `taint` from `k/k` to `k8s.io/cloud-provider`

### PR DESCRIPTION
**Pull request by stewart-yu**
**What type of PR is this?**
 /kind feature
/kind cleanup

**What this PR does / why we need it**:
In PR #73491, we using a copy function `addOrUpdateTaint` in 'k8s.io/cloud-provider', which have public function `AddOrUpdateTaint` in `k/k` yet.
In this PR, we want to remove the duplication. 

1. The first part, move logic about `taint` from `k/k` to `k8s.io/cloud-provider`;
2. The secondary part, move the logic about `node` handler in `k/k` to `k8s.io/cloud-provider/node/testutil`, so we can move some test file about `taint`  to `k8s.io/cloud-provider`;
3. The Third part, move test file about `taint` from `k/k` to `k8s.io/cloud-provider`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go
Comment (yue9944882) [inline staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go]: this repo is for universal apimachinery pkgs. we dont want any `k8s.io/api` dependency here

**Review by yue9944882 — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go
Comment (stewart-yu) [reply staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go]: little knowledge about it. Do you have any suggestion about where to put it?  how about `vendor/k8s.io/util`?

Comment (yue9944882) [reply staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go]: > how about vendor/k8s.io/util?

mmmmmmm...idk @sttts any thought?

**Comment by andrewsykim** [comment]
ref https://github.com/kubernetes/kubernetes/issues/69585

**Comment by jennybuckley** [comment]
/assign @caesarxuchao 
cc @mbohlool

**Comment by caesarxuchao** [comment]
Taint is a specific API, it doesn't belong to apimachinery.

I don't think it belongs to k8s/api either, as we want to keep k8s/api only about the api definition.

@sttts @mengqiy shall we create a repo for Kubernetes API helpers?

**Comment by mengqiy** [comment]
I agree it doesn't belong to apimachinery.
But I found that there are a few API helper in `k8s.io/api` repo:
- https://github.com/kubernetes/api/blob/master/core/v1/taint.go
- https://github.com/kubernetes/api/blob/master/core/v1/toleration.go

**Comment by caesarxuchao** [comment]
> But I found that there are a few API helper in k8s.io/api repo:

I don't think they are intentional.

**Comment by yue9944882** [comment]
> I agree it doesn't belong to apimachinery.
> But I found that there are a few API helper in k8s.io/api repo:

yuh huh. matter of fact, we usually put these helper functions under their types definition pkg named `helpers.go`. 

> I don't think they are intentional.

then i'll worry about circular dependencies.

File: staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go
Comment (stewart-yu) [reply staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go]: In order to remove `pkg/controller` dependencies from `k8s.io/kubernetes/pkg/cloudprovider/providers`, see https://github.com/kubernetes/kubernetes/pull/70005/files#r226647860, function `AddOrUpdateTaintOnNode()` https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L903 must move out `k/k`, but innner function https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L927 depend on `k8s.io/kubernetes/pkg/util/taints`. So this PR done something about it

**Comment by andrewsykim** [comment]
Any objections to move pkg/util/taints to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/taint.go? It's not ideal but it will unblock the out-of-tree cloud provider efforts and it seems we've already put some taint utility methods there.

File: staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go
Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go]: We don't have a place for helpers like this. What would happen if we move them into k8s.io/cloud-provider? We could still use them from k/k code.

Comment (andrewsykim) [reply staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go]: `k8s.io/cloud-provider` works for me! It might be better to just duplicate the 1 method we need rather than moving this whole package over there. Once there's a better place for helpers like this we can remove the duplicate code. It's a bit hacky but it'll unblock us for the time being.

Comment (andrewsykim) [reply staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go]: https://github.com/kubernetes/kubernetes/pull/73491 is going to unblock us for now. Would be great if this PR can refactor things a bit so that taint util doesn't have to be duplicated in k8s.io/cloud-provider :)

Comment (stewart-yu) [reply staging/src/k8s.io/apimachinery/pkg/util/taints/taints.go]: yes, great work :+1:

**Comment by dims** [comment]
@stewart-yu can you please rebase? let's try to get this into 1.14?

**Comment by stewart-yu** [comment]
ok, will do it ASAP

**Comment by andrewsykim** [comment]
@stewart-yu might be good to get someone from SIG apimachinery to approve the new package location for that. There seems to be different opinions on where the taint functions should live.

**Comment by stewart-yu** [comment]
@kubernetes/sig-api-machinery-pr-reviews

**Comment by dims** [comment]
@stewart-yu can you please rebase?

**Comment by soggiest** [comment]
Hello!
1.14 code freeze is coming in about 3 days, will this be rebased and merged in the next week?

**Comment by spiffxp** [comment]
/hold
v1.14 release lead here, this is a size/XXL PR that is attempting to land right before Code Freeze... are you _really sure_ this is a good idea?

**Comment by vllry** [comment]
Ping

**Comment by andrewsykim** [comment]
/close 

I'm closing this for now, I think this can be useful for https://github.com/kubernetes/kubernetes/issues/81172 but let's reopen when we get there.

**Pull request closed without merging**
